File: backend/tests/upload.test.js
const request = require('supertest');
const fs = require('fs');
const os = require('os');
const path = require('path');

jest.mock('../src/config/database', () => ({
  query: jest.fn(),
  connect: jest.fn(),
  on: jest.fn(),
}));

// Mock imageService: simply return the input path unchanged (no actual resize)
jest.mock('../src/services/imageService', () => ({
  resizePhoto: jest.fn().mockImplementation((inputPath) => Promise.resolve(inputPath)),
}));

const app = require('../src/app');
const pool = require('../src/config/database');

// Minimal 1×1 white JPEG (from a known-good tiny JPEG buffer)
const TINY_JPEG = Buffer.from(
  '/9j/4AAQSkZJRgABAQEASABIAAD/2wBDAAgGBgcGBQgHBwcJCQgKDBQNDAsLDBkSEw8U' +
  'HRofHh0aHBwgJC4nICIsIxwcKDcpLDAxNDQ0Hyc5PTgyPC4zNDL/wAARAQABAAEDASIA' +
  'AhEBAxEB/8QAFAABAAAAAAAAAAAAAAAAAAAACf/EABQQAQAAAAAAAAAAAAAAAAAAAAD/' +
  'xAAUAQEAAAAAAAAAAAAAAAAAAAAA/8QAFBEBAAAAAAAAAAAAAAAAAAAAAP/aAAwDAQACEQ' +
  'MRADwAKwAB/9k=',
  'base64'
);

describe('Photo upload', () => {
  beforeEach(() => jest.clearAllMocks());

  it('returns 400 when no file is provided', async () => {
    const res = await request(app).post('/api/students/1/photo');
    expect(res.status).toBe(400);
  });

  it('returns 404 when student does not exist', async () => {
    pool.query.mockResolvedValue({ rows: [] }); // Student.findById returns nothing

    const res = await request(app)
      .post('/api/students/999/photo')
      .attach('photo', TINY_JPEG, { filename: 'test.jpg', contentType: 'image/jpeg' });

    expect(res.status).toBe(404);
  });

  it('resizes the photo and updates the student', async () => {
    pool.query
      .mockResolvedValueOnce({ rows: [{ id: 1, first_name: 'Jean', last_name: 'Dupont' }] }) // findById
      .mockResolvedValueOnce({
        rows: [{ id: 1, first_name: 'Jean', last_name: 'Dupont', photo_url: 'uploads/photo.jpg' }],
      }); // updatePhoto

    const res = await request(app)
      .post('/api/students/1/photo')
      .attach('photo', TINY_JPEG, { filename: 'test.jpg', contentType: 'image/jpeg' });

    expect(res.status).toBe(200);
    expect(res.body.photo_url).toBeDefined();

    const { resizePhoto } = require('../src/services/imageService');
    expect(resizePhoto).toHaveBeenCalled();
  });

  it('rejects files that are not images', async () => {
    const res = await request(app)
      .post('/api/students/1/photo')
      .attach('photo', Buffer.from('not an image'), {
        filename: 'evil.pdf',
        contentType: 'application/pdf',
      });

    // Multer rejects non-images with 415 or 500 depending on the error
    expect([400, 415, 500]).toContain(res.status);
  });
});

File: backend/tests/upload_test.py
from unittest.mock import patch, MagicMock
from fastapi.testclient import TestClient

from src.main import app

client = TestClient(app)

# Minimal valid JPEG header (FFD8FFE0 + padding)
TINY_JPEG = bytes([0xFF, 0xD8, 0xFF, 0xE0]) + b"\x00" * 16


class TestPhotoUpload:
    def test_returns_415_when_file_type_is_not_image(self):
        res = client.post(
            "/api/students/1/photo",
            files={"photo": ("evil.pdf", b"not an image", "application/pdf")},
        )
        assert res.status_code == 415

    def test_returns_404_when_student_does_not_exist(self):
        with patch("src.routers.students.StudentModel.find_by_id") as mock_find, \
             patch("src.routers.students.resize_photo") as mock_resize, \
             patch("builtins.open", MagicMock()):
            mock_find.return_value = None
            mock_resize.return_value = "/tmp/photo.jpg"
            res = client.post(
                "/api/students/999/photo",
                files={"photo": ("test.jpg", TINY_JPEG, "image/jpeg")},
            )
        assert res.status_code == 404

    def test_resizes_photo_and_updates_student(self, tmp_path):
        fake_photo = tmp_path / "photo.jpg"
        fake_photo.write_bytes(b"fake image data")

        with patch("src.routers.students.StudentModel.find_by_id") as mock_find, \
             patch("src.routers.students.StudentModel.update_photo") as mock_update, \
             patch("src.routers.students.resize_photo") as mock_resize, \
             patch("src.routers.students.UPLOAD_DIR", str(tmp_path)):
            mock_find.return_value = {"id": 1, "first_name": "Jean", "last_name": "Dupont"}
            mock_update.return_value = {
                "id": 1, "first_name": "Jean", "last_name": "Dupont", "photo_url": "uploads/photo.jpg"
            }
            mock_resize.return_value = str(fake_photo)
            res = client.post(
                "/api/students/1/photo",
                files={"photo": ("test.jpg", TINY_JPEG, "image/jpeg")},
            )
        assert res.status_code == 200
        assert res.json()["photo_url"] is not None
        mock_resize.assert_called_once()

    def test_rejects_non_image_files(self):
        res = client.post(
            "/api/students/1/photo",
            files={"photo": ("evil.pdf", b"not an image", "application/pdf")},
        )
        assert res.status_code in (400, 415, 422)
